File: src/eval/preference-dataset.ts
import type { SpellObservation } from "../domain/spell.ts";
import {
  AGENT_GYM_FAMILY_SPLIT_SIZES,
  type AgentGymFamilyId,
  type AgentGymSplit,
} from "../scenarios/agent-gym-family.ts";
import { createSpellToolManifest } from "../tools/definitions.ts";
import type { AgentGymSnapshot, AgentGymStep } from "./agent-gym.ts";
import {
  AGENT_GYM_POLICY_BASELINES,
  runAgentGymPolicy,
  type AgentGymPolicyId,
} from "./policy-benchmark.ts";

export const AGENT_GYM_PREFERENCE_SCHEMA = "hex-machina-agent-gym-preference-group/v2" as const;
export const AGENT_GYM_PREFERENCE_VERIFIER_PROTOCOL = "hex-machina-agent-gym-preference-verifier/v2" as const;
export const MAX_PREFERENCE_GROUPS = 256;

function pairCount(candidateCount: number) {
  return candidateCount * (candidateCount - 1) / 2;
}

function stableDecimal(value: number) {
  return Math.round(value * 1_000_000) / 1_000_000;
}

export interface AgentGymPreferenceCandidate {
  policyId: AgentGymPolicyId;
  rank: number;
  reward: number;
  advantage: number;
  status: AgentGymSnapshot["status"];
  terminationReason: AgentGymSnapshot["terminationReason"];
  unsafeMutation: boolean;
  constraintViolation: boolean;
  constraintPreserved: boolean | null;
  invalidActionCount: number;
  actionCount: number;
  transitions: AgentGymStep[];
}

export interface AgentGymPreferenceGroup {
  schema: typeof AGENT_GYM_PREFERENCE_SCHEMA;
  environmentProtocol: "hex-machina-agent-gym/v1";
  observationSchema: "hex-machina-public-spell-graph/v1";
  actionManifest: ReturnType<typeof createSpellToolManifest>;
  familyId: AgentGymFamilyId;
  scenarioId: string;
  split: AgentGymSplit;
  variantIndex: number;
  seed: number;
  task: AgentGymSnapshot["task"];
  initialObservation: SpellObservation;
  initialStateKey: string;
  maxScore: number;
  groupMeanReward: number;
  candidates: AgentGymPreferenceCandidate[];
  preferencePairs: Array<{
    chosenPolicyId: AgentGymPolicyId;
    rejectedPolicyId: AgentGymPolicyId;
    rewardMargin: number;
  }>;
}

export interface PreferenceVerificationIssue {
  line: number;
  scenarioId?: string;
  code: "invalid-json" | "invalid-record" | "duplicate-scenario" | "group-mismatch";
  message: string;
}

function recordOf(value: unknown): Record<string, unknown> | null {
  return typeof value === "object" && value !== null && !Array.isArray(value)
    ? value as Record<string, unknown>
    : null;
}

function stableValue(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(stableValue);
  const record = recordOf(value);
  if (!record) return value;
  return Object.fromEntries(Object.keys(record).sort().map((key) => [key, stableValue(record[key])]));
}

function stableJson(value: unknown) {
  return JSON.stringify(stableValue(value));
}

function candidateFromSnapshot(
  policyId: AgentGymPolicyId,
  snapshot: AgentGymSnapshot,
): Omit<AgentGymPreferenceCandidate, "rank" | "advantage"> {
  return {
    policyId,
    reward: snapshot.score,
    status: snapshot.status,
    terminationReason: snapshot.terminationReason,
    unsafeMutation: snapshot.trajectory.some((step) => (
      step.rewardReasons.some((reason) => reason.includes("before explaining"))
    )),
    constraintViolation: snapshot.terminationReason === "constraint-violated",
    constraintPreserved: snapshot.constraintPreserved,
    invalidActionCount: snapshot.trajectory.filter((step) => step.error !== undefined).length,
    actionCount: snapshot.trajectory.length,
    transitions: snapshot.trajectory,
  };
}

export async function buildAgentGymPreferenceGroup(options: {
  family: AgentGymFamilyId;
  split: AgentGymSplit;
  index: number;
}): Promise<AgentGymPreferenceGroup> {
  const policyIds = AGENT_GYM_POLICY_BASELINES.map((baseline) => baseline.id);
  const snapshots = await Promise.all(policyIds.map((policyId) => runAgentGymPolicy(policyId, options)));
  const reference = snapshots[0];
  if (!snapshots.every((snapshot) => (
    snapshot.scenarioId === reference.scenarioId &&
    stableJson(snapshot.task) === stableJson(reference.task) &&
    stableJson(snapshot.initialObservation) === stableJson(reference.initialObservation) &&
    snapshot.initialStateKey === reference.initialStateKey
  ))) {
    throw new Error("Preference candidates do not share one deterministic reset context");
  }

  const projected = snapshots.map((snapshot, index) => (
    candidateFromSnapshot(policyIds[index], snapshot)
  ));
  const groupMeanReward = stableDecimal(
    projected.reduce((total, candidate) => total + candidate.reward, 0) / projected.length,
  );
  const candidates = projected
    .sort((left, right) => right.reward - left.reward || left.policyId.localeCompare(right.policyId))
    .map((candidate, index) => ({
      ...candidate,
      rank: index + 1,
      advantage: stableDecimal(candidate.reward - groupMeanReward),
    }));
  const preferencePairs = candidates.flatMap((chosen, chosenIndex) => (
    candidates.slice(chosenIndex + 1).map((rejected) => ({
      chosenPolicyId: chosen.policyId,
      rejectedPolicyId: rejected.policyId,
      rewardMargin: chosen.reward - rejected.reward,
    }))
  ));

  return {
    schema: AGENT_GYM_PREFERENCE_SCHEMA,
    environmentProtocol: reference.protocol,
    observationSchema: "hex-machina-public-spell-graph/v1",
    actionManifest: createSpellToolManifest(),
    familyId: reference.familyId,
    scenarioId: reference.scenarioId,
    split: options.split,
    variantIndex: options.index,
    seed: reference.seed,
    task: reference.task,
    initialObservation: reference.initialObservation,
    initialStateKey: reference.initialStateKey,
    maxScore: reference.maxScore,
    groupMeanReward,
    candidates,
    preferencePairs,
  };
}

export async function collectAgentGymPreferenceGroups(split: AgentGymSplit = "train") {
  const groups: AgentGymPreferenceGroup[] = [];
  for (const family of Object.keys(AGENT_GYM_FAMILY_SPLIT_SIZES) as AgentGymFamilyId[]) {
    for (let index = 0; index < AGENT_GYM_FAMILY_SPLIT_SIZES[family][split]; index += 1) {
      groups.push(await buildAgentGymPreferenceGroup({ family, split, index }));
    }
  }
  return groups;
}

export function serializeAgentGymPreferenceGroupsJsonl(groups: AgentGymPreferenceGroup[]) {
  return `${groups.map((group) => JSON.stringify(group)).join("\n")}\n`;
}

function parseGroup(value: unknown): AgentGymPreferenceGroup | string {
  const record = recordOf(value);
  if (!record) return "Preference group must be a JSON object";
  if (record.schema !== AGENT_GYM_PREFERENCE_SCHEMA) return `schema must be ${AGENT_GYM_PREFERENCE_SCHEMA}`;
  if (typeof record.familyId !== "string" || !Object.hasOwn(AGENT_GYM_FAMILY_SPLIT_SIZES, record.familyId)) {
    return "familyId is unknown";
  }
  if (typeof record.split !== "string" || !["train", "validation", "test"].includes(record.split)) {
    return "split must be train, validation, or test";
  }
  const familyId = record.familyId as AgentGymFamilyId;
  const split = record.split as AgentGymSplit;
  if (!Number.isInteger(record.variantIndex) || (record.variantIndex as number) < 0 ||
      (record.variantIndex as number) >= AGENT_GYM_FAMILY_SPLIT_SIZES[familyId][split]) {
    return "variantIndex is outside the selected family split";
  }
  const candidateCount = AGENT_GYM_POLICY_BASELINES.length;
  if (typeof record.scenarioId !== "string" || !Array.isArray(record.candidates) ||
      record.candidates.length !== candidateCount ||
      !Array.isArray(record.preferencePairs) || record.preferencePairs.length !== pairCount(candidateCount)) {
    return `scenarioId, ${candidateCount} candidates, and ${pairCount(candidateCount)} preference pairs are required`;
  }
  return record as unknown as AgentGymPreferenceGroup;
}

export async function verifyAgentGymPreferenceGroupsJsonl(jsonl: string) {
  const lines = jsonl.split(/\r?\n/).map((line) => line.trim()).filter(Boolean);
  const issues: PreferenceVerificationIssue[] = [];
  if (lines.length === 0) {
    issues.push({ line: 0, code: "invalid-record", message: "Dataset contains no preference groups" });
  }
  if (lines.length > MAX_PREFERENCE_GROUPS) {
    issues.push({
      line: 0,
      code: "invalid-record",
      message: `Dataset exceeds the ${MAX_PREFERENCE_GROUPS}-group verification limit`,
    });
  }

  let verifiedGroups = 0;
  const seenScenarios = new Set<string>();
  for (const [lineIndex, line] of lines.slice(0, MAX_PREFERENCE_GROUPS).entries()) {
    const lineNumber = lineIndex + 1;
    let decoded: unknown;
    try {
      decoded = JSON.parse(line);
    } catch {
      issues.push({ line: lineNumber, code: "invalid-json", message: "Line is not valid JSON" });
      continue;
    }
    const parsed = parseGroup(decoded);
    if (typeof parsed === "string") {
      issues.push({ line: lineNumber, code: "invalid-record", message: parsed });
      continue;
    }
    if (seenScenarios.has(parsed.scenarioId)) {
      issues.push({
        line: lineNumber,
        scenarioId: parsed.scenarioId,
        code: "duplicate-scenario",
        message: "Scenario appears more than once",
      });
      continue;
    }
    seenScenarios.add(parsed.scenarioId);

    const expected = await buildAgentGymPreferenceGroup({
      family: parsed.familyId,
      split: parsed.split,
      index: parsed.variantIndex,
    });
    if (stableJson(expected) !== stableJson(parsed)) {
      issues.push({
        line: lineNumber,
        scenarioId: parsed.scenarioId,
        code: "group-mismatch",
        message: "Preference group differs from deterministic policy regeneration",
      });
      continue;
    }
    verifiedGroups += 1;
  }

  return {
    protocol: AGENT_GYM_PREFERENCE_VERIFIER_PROTOCOL,
    valid: issues.length === 0,
    groupCount: lines.length,
    verifiedGroups,
    issueCount: issues.length,
    issues,
  };
}

File: adapters/preference_dataset.py
"""Streaming Python access to verified Hexmend preference groups."""

from __future__ import annotations

import json
import subprocess
from pathlib import Path
from typing import Any, Iterator, Mapping, Sequence


PREFERENCE_GROUP_SCHEMA = "hexmend-agent-gym-preference-group/v2"
PREFERENCE_PAIR_SCHEMA = "hexmend-agent-gym-preference-pair/v2"
PREFERENCE_VERIFIER_PROTOCOL = "hexmend-agent-gym-preference-verifier/v2"
EXPECTED_POLICY_IDS = {
    "grounded-reference",
    "mutate-before-explain",
    "diagnosis-only",
    "constraint-violating",
    "memorized-canonical-ids",
}
MAX_DATASET_BYTES = 64 * 1024 * 1024
MAX_GROUP_BYTES = 4 * 1024 * 1024
MAX_GROUPS = 256
PAIR_CONTEXT_FIELDS = (
    "environmentProtocol",
    "observationSchema",
    "actionManifest",
    "familyId",
    "scenarioId",
    "split",
    "variantIndex",
    "seed",
    "task",
    "initialObservation",
    "initialStateKey",
)


class HexmendPreferenceError(ValueError):
    """Raised when a preference artifact violates its versioned contract."""


def _record(value: Any, label: str) -> Mapping[str, Any]:
    if not isinstance(value, dict):
        raise HexmendPreferenceError(f"{label} must be a JSON object")
    return value


def _number(value: Any, label: str) -> float:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        raise HexmendPreferenceError(f"{label} must be numeric")
    return float(value)


def _optional_filter(value: str | None, label: str) -> str | None:
    if value is None:
        return None
    if not isinstance(value, str) or not value.strip():
        raise HexmendPreferenceError(f"{label} filter must be a non-empty string")
    return value


def _validate_group(value: Any, line_number: int) -> Mapping[str, Any]:
    prefix = f"preference group on line {line_number}"
    group = _record(value, prefix)
    if group.get("schema") != PREFERENCE_GROUP_SCHEMA:
        raise HexmendPreferenceError(f"{prefix} has an unsupported schema")
    if not isinstance(group.get("scenarioId"), str) or not group["scenarioId"]:
        raise HexmendPreferenceError(f"{prefix} requires scenarioId")
    if not isinstance(group.get("familyId"), str) or not group["familyId"]:
        raise HexmendPreferenceError(f"{prefix} requires familyId")
    if not isinstance(group.get("split"), str) or not group["split"]:
        raise HexmendPreferenceError(f"{prefix} requires split")
    if any(field not in group for field in PAIR_CONTEXT_FIELDS):
        raise HexmendPreferenceError(f"{prefix} is missing trainer context")

    candidates = group.get("candidates")
    pairs = group.get("preferencePairs")
    if not isinstance(candidates, list) or len(candidates) != 5:
        raise HexmendPreferenceError(f"{prefix} must contain five candidates")
    if not isinstance(pairs, list) or len(pairs) != 10:
        raise HexmendPreferenceError(f"{prefix} must contain ten preference pairs")

    candidate_by_policy: dict[str, Mapping[str, Any]] = {}
    for expected_rank, candidate_value in enumerate(candidates, start=1):
        candidate = _record(candidate_value, f"candidate {expected_rank} on line {line_number}")
        policy_id = candidate.get("policyId")
        if not isinstance(policy_id, str) or policy_id not in EXPECTED_POLICY_IDS or policy_id in candidate_by_policy:
            raise HexmendPreferenceError(f"{prefix} contains an unknown or duplicate policy")
        if candidate.get("rank") != expected_rank:
            raise HexmendPreferenceError(f"{prefix} candidate ranks are not contiguous")
        _number(candidate.get("reward"), f"{prefix} candidate reward")
        _number(candidate.get("advantage"), f"{prefix} candidate advantage")
        if not isinstance(candidate.get("transitions"), list):
            raise HexmendPreferenceError(f"{prefix} candidate transitions must be an array")
        termination_reason = candidate.get("terminationReason")
        expected_violation = termination_reason == "constraint-violated"
        expected_preserved = (
            False if expected_violation else True if termination_reason == "goal-verified" else None
        )
        if candidate.get("constraintViolation") is not expected_violation:
            raise HexmendPreferenceError(f"{prefix} contains an inconsistent constraint violation label")
        if candidate.get("constraintPreserved") is not expected_preserved:
            raise HexmendPreferenceError(f"{prefix} contains an inconsistent constraint preservation label")
        candidate_by_policy[policy_id] = candidate

    if set(candidate_by_policy) != EXPECTED_POLICY_IDS:
        raise HexmendPreferenceError(f"{prefix} does not contain the expected policy controls")
    rewards = [_number(candidate["reward"], f"{prefix} candidate reward") for candidate in candidates]
    if any(left <= right for left, right in zip(rewards, rewards[1:])):
        raise HexmendPreferenceError(f"{prefix} rewards are not strictly descending")
    mean_reward = sum(rewards) / len(rewards)
    if abs(_number(group.get("groupMeanReward"), f"{prefix} group mean") - mean_reward) > 1e-9:
        raise HexmendPreferenceError(f"{prefix} group mean does not match its rewards")
    for candidate in candidates:
        if abs(_number(candidate["advantage"], f"{prefix} candidate advantage") - (_number(candidate["reward"], f"{prefix} candidate reward") - mean_reward)) > 1e-9:
            raise HexmendPreferenceError(f"{prefix} contains an inconsistent centered advantage")

    expected_pairs = {
        (candidates[chosen]["policyId"], candidates[rejected]["policyId"])
        for chosen in range(len(candidates))
        for rejected in range(chosen + 1, len(candidates))
    }
    observed_pairs: set[tuple[str, str]] = set()
    for pair_value in pairs:
        pair = _record(pair_value, f"preference pair on line {line_number}")
        chosen_id = pair.get("chosenPolicyId")
        rejected_id = pair.get("rejectedPolicyId")
        if not isinstance(chosen_id, str) or not isinstance(rejected_id, str):
            raise HexmendPreferenceError(f"{prefix} pair policy IDs must be strings")
        identity = (chosen_id, rejected_id)
        if identity not in expected_pairs or identity in observed_pairs:
            raise HexmendPreferenceError(f"{prefix} contains an unknown or duplicate pair")
        expected_margin = (
            float(candidate_by_policy[chosen_id]["reward"])
            - float(candidate_by_policy[rejected_id]["reward"])
        )
        if _number(pair.get("rewardMargin"), f"{prefix} preference margin") != expected_margin or expected_margin <= 0:
            raise HexmendPreferenceError(f"{prefix} contains an inconsistent preference margin")
        observed_pairs.add(identity)
    if observed_pairs != expected_pairs:
        raise HexmendPreferenceError(f"{prefix} does not contain every ranked pair")
    return group


class HexmendPreferenceDataset:
    """Bounded, repeatable streaming view over preference-group JSONL.

    Call :meth:`verify` before training to rerun the canonical TypeScript
    verifier. Calls to :meth:`groups` and :meth:`pairs` then stream one JSONL
    group at a time, so pair iteration never loads the complete corpus.
    """

    def __init__(self, path: str | Path) -> None:
        self.path = Path(path)

    def _validate_size(self) -> None:
        size = self.path.stat().st_size
        if size == 0:
            raise HexmendPreferenceError("preference dataset is empty")
        if size > MAX_DATASET_BYTES:
            raise HexmendPreferenceError("preference dataset exceeds the 64 MiB limit")

    def verify(
        self,
        command: Sequence[str] | None = None,
        cwd: str | Path | None = None,
    ) -> Mapping[str, Any]:
        """Regenerate every labeled policy with the production verifier."""

        self._validate_size()
        repository = Path(__file__).resolve().parents[1]
        selected_command = list(command or (
            str(repository / "node_modules" / ".bin" / "tsx"),
            "scripts/verify-agent-gym-preferences.ts",
        ))
        with self.path.open("r", encoding="utf-8") as source:
            result = subprocess.run(
                selected_command,
                cwd=Path(cwd) if cwd is not None else repository,
                stdin=source,
                capture_output=True,
                text=True,
                check=False,
            )
        try:
            receipt = json.loads(result.stdout)
        except json.JSONDecodeError as error:
            detail = result.stderr.strip() or "verifier returned no JSON receipt"
            raise HexmendPreferenceError(detail) from error
        if result.returncode != 0 or receipt.get("valid") is not True:
            issue = receipt.get("issues", [{}])[0].get("message", "verification failed")
            raise HexmendPreferenceError(str(issue))
        if receipt.get("protocol") != PREFERENCE_VERIFIER_PROTOCOL:
            raise HexmendPreferenceError("verifier returned an unsupported protocol")
        return receipt

    def groups(
        self,
        *,
        split: str | None = None,
        family: str | None = None,
    ) -> Iterator[Mapping[str, Any]]:
        """Yield validated groups in file order, optionally restricting a curriculum."""

        self._validate_size()
        selected_split = _optional_filter(split, "split")
        selected_family = _optional_filter(family, "family")
        seen_scenarios: set[str] = set()
        group_count = 0
        matched_count = 0
        with self.path.open("r", encoding="utf-8") as source:
            for line_number, line in enumerate(source, start=1):
                if not line.strip():
                    continue
                if len(line.encode("utf-8")) > MAX_GROUP_BYTES:
                    raise HexmendPreferenceError(
                        f"preference group on line {line_number} exceeds the 4 MiB limit"
                    )
                group_count += 1
                if group_count > MAX_GROUPS:
                    raise HexmendPreferenceError("preference dataset exceeds 256 groups")
                try:
                    decoded = json.loads(line)
                except json.JSONDecodeError as error:
                    raise HexmendPreferenceError(
                        f"preference group on line {line_number} is not valid JSON"
                    ) from error
                group = _validate_group(decoded, line_number)
                scenario_id = str(group["scenarioId"])
                if scenario_id in seen_scenarios:
                    raise HexmendPreferenceError(f"duplicate scenarioId {scenario_id}")
                seen_scenarios.add(scenario_id)
                if selected_split is not None and group["split"] != selected_split:
                    continue
                if selected_family is not None and group["familyId"] != selected_family:
                    continue
                matched_count += 1
                yield group
        if group_count == 0:
            raise HexmendPreferenceError("preference dataset is empty")
        if matched_count == 0:
            filters = ", ".join(
                value
                for value in (
                    f"split={selected_split!r}" if selected_split is not None else "",
                    f"family={selected_family!r}" if selected_family is not None else "",
                )
                if value
            )
            raise HexmendPreferenceError(f"no preference groups match {filters}")

    def pairs(
        self,
        *,
        split: str | None = None,
        family: str | None = None,
    ) -> Iterator[Mapping[str, Any]]:
        """Project a filtered group stream into chosen/rejected records."""

        for group in self.groups(split=split, family=family):
            candidate_by_policy = {
                candidate["policyId"]: candidate for candidate in group["candidates"]
            }
            for pair in group["preferencePairs"]:
                yield {
                    "schema": PREFERENCE_PAIR_SCHEMA,
                    **{field: group[field] for field in PAIR_CONTEXT_FIELDS},
                    "chosen": candidate_by_policy[pair["chosenPolicyId"]],
                    "rejected": candidate_by_policy[pair["rejectedPolicyId"]],
                    "rewardMargin": pair["rewardMargin"],
                }
